include/seqan3/range/container/bitcompressed_vector.hpp

```diff
@@ -0,0 +1,1114 @@
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides seqan3::bitcompressed_vector.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <type_traits>
+
+#include <range/v3/iterator_range.hpp>
+
+#include <sdsl/int_vector.hpp>
+
+#include <seqan3/core/concept/cereal.hpp>
+#include <seqan3/core/metafunction/range.hpp>
+#include <seqan3/range/container/concept.hpp>
+#include <seqan3/range/detail/random_access_iterator.hpp>
+#include <seqan3/range/view/to_rank.hpp>
+#include <seqan3/range/view/convert.hpp>
+#include <seqan3/std/concept/iterator.hpp>
+
+namespace seqan3
+{
+
+/*!\brief A space-optimised version of std::vector that compresses multiple letters into a single byte.
+ * \tparam alphabet_type The value type of the container, must satisfy seqan3::alphabet_concept and not be `const` or
+ *                       `&`.
+ * \implements seqan3::reservable_container_concept
+ * \implements seqan3::cerealisable_concept
+ * \ingroup container
+ *
+ * This class template behaves just like std::vector<alphabet_type> but has an internal representation where
+ * multiple values are packed into a single byte/word to save space, e.g. bitcompressed_vector<seqan3::dna4> uses a
+ * quarter of of the memory that std::vector<seqan3::dna4> uses, because a single seqan3::dna4 letter can be represented
+ * in two bits (instead of 8 which is the lower bound for a single object in C++.
+ *
+ * The disadvantages are slightly slower operations and unsafety towards parallel writes to adjacent positions
+ * in the bitcompressed_vector.
+ *
+ *
+ * ### Example
+ *
+ * ```cpp
+ * std::vector<dna4>          v0{"ACGT"_dna4}; // occupies 4 bytes in memory
+ * bitcompressed_vector<dna4> v1{"ACGT"_dna4}; // occupies 1 byte in memory
```

Comment:
Yes and no. I would consider that an edge effect of a very small container. It also not true that it occupies 4 bytes, because it also needs to store a pointer. Also because of amortised space allocation it likely already reserved at least 32bytes in the target.... *but the point is* that the amortised space consumption in regard to the size n is not n, but n/4.
Writing anything else would be really confusing, or not?